[Refactoring] Improve Void handling for async conversion

hamishknight opened the pull request:
When converting a function with a completion handler that has a Void success parameter, e.g `(Void?, Error?) -> Void`, or more likely a `Result<Void, Error>` parameter, make sure to omit the `-> Void` from the resulting async function conversion.

In addition, strip any Void bindings from an async function call, and any explicit Void return values from inside the async function.

Resolves rdar://75189289

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4911,42 +4972,24 @@ class AsyncConverter : private SourceEntityWalker {
-    }
-
-    if (HandlerParams.empty()) {
+    SmallVector<Type, 2> scratch;
```
Same here

bnbarham commented on lib/IDE/Refactoring.cpp:
```diff
@@ -4187,12 +4198,62 @@ struct AsyncHandlerDesc {
+
+  /// Gets the return value types for the async equivalent of this handler.
+  ArrayRef<swift::Type>
+  getAsyncReturnTypes(SmallVectorImpl<swift::Type> &scratch) const {
```
Scratch to be consistent? (as much as I hate this naming style, it is what the rest of the file was 😞)

hamishknight replied:
Ah good catch! Will fix that

bnbarham reviewed (approved):
Other than the style nitpick, this looks great! Thanks for the nice refactors, the param handling is much nicer now.

hamishknight committed:
[Refactoring] Improve Void handling for async conversion

When converting a function with a completion handler
that has a Void success parameter, e.g
`(Void?, Error?) -> Void`, or more likely a
`Result<Void, Error>` parameter, make sure to omit
the `-> Void` from the resulting async function
conversion.

In addition, strip any Void bindings from an async
function call, and any explicit Void return values
from inside the async function.

Resolves rdar://75189289

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
Windows failure is unrelated
